OSDOCS-19006: Add Azure procedure for migrating x86 control plane to arm64 (#112154)
openshift/openshift-docs: modules/migrating-from-x86-to-arm-cp-azure.adoc

// Module included in the following assemblies:
//
// * updating/updating_a_cluster/migrating-to-multi-payload.adoc

:_mod-docs-content-type: PROCEDURE
[id="migrating-from-x86-to-arm64-cp-azure_{context}"]
= Migrating the x86 control plane to arm64 architecture on {azure-full}

[role="_abstract"]

> You can migrate the control plane in your cluster from `x86` to `arm64` architecture on {azure-first}.

Reviewer: Ideally, we'd say a little something about why the user wants to do this. What is their goal/the underlying user story here that motivates them to migrate the control plane in the cluster from `x86` to `arm64` architecture

> Unlike {aws-short}, which provides pre-published `arm64` Amazon Machine Images (AMIs), {azure-short} requires you to manually create a gallery image from the `arm64` {op-system-first} VHD before updating the control plane machine set.

Reviewer: We shouldn't compare providers like this, the procedure should stand on its own for the Azure user
```suggestion
{azure-short} requires you to manually create a gallery image from the `arm64` {op-system-first} VHD before updating the control plane machine set.
```

worker-001.example.com        Ready    worker                 100d   v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.x86_64    cri-o://1.30.x
worker-002.example.com        Ready    worker                 98d    v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.x86_64    cri-o://1.30.x
master-001.example.com        Ready    control-plane,master   120d   v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.x86_64    cri-o://1.30.x
master-002.example.com        Ready    control-plane,master   120d   v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.x86_64    cri-o://1.30.x
master-003.example.com        Ready    control-plane,master   120d   v1.30.7   10.x.x.x    <none>        Red Hat Enterprise Linux CoreOS 4xx.xx.xxxxx-0   5.x.x-xxx.x.x.el9_xx.x86_64    cri-o://1.30.x
----
+
The `KERNEL-VERSION` field in the output indicates the architecture of the nodes.

> . Check that your cluster uses the multi payload by running the following command:

Reviewer: ```suggestion
. Check that your cluster is multi-architecture compatible by running the following command:
```

+
[source,terminal]
----
$ oc adm release info -o jsonpath="{ .metadata.metadata}"
----
+
If you see the following output, the cluster is multi-architecture compatible.
+
[source,terminal]
----
{
 "release.openshift.io/architecture": "multi",
 "url": "https://access.redhat.com/errata/<errata_version>"
}
----
+

> If the cluster is not using the multi payload, migrate the cluster to a multi-architecture cluster. For more information, see "Migrating to a cluster with multi-architecture compute machines using the CLI".

Reviewer: ```suggestion
If the value of the `release.openshift.io/architecture` field is not `multi`, migrate the cluster to a multi-architecture cluster. For more information, see "Migrating to a cluster with multi-architecture compute machines using the CLI".
```

. Update your image stream from single-architecture to multi-architecture by running the following command:
+
--
include::snippets/update-image-stream-to-multi-arch.adoc[]
--

> . Set environment variables for the cluster's {azure-short} resources by running the following commands:

Reviewer: ```suggestion
. Set environment variables for the {azure-short} resources by running the following commands:
```

+
[source,terminal]
----
$ INFRA_ID=$(oc get infrastructure cluster -o jsonpath='{.status.infrastructureName}')
$ REGION=$(oc get infrastructure cluster -o jsonpath='{.status.platformStatus.azure.cloudName}' && oc get machines -n openshift-machine-api -o jsonpath='{.items[0].spec.providerSpec.value.location}')
$ RESOURCE_GROUP="${INFRA_ID}-rg"

> ----

Reviewer: We require one command only per block, you can leave this under a single step but please break into three `[source,terminal]` blocks with only one each

. Get the cluster's storage account name and shared image gallery name by running the following commands:
+
[source,terminal]
----
$ STORAGE_ACCOUNT_NAME=$(az storage account list --resource-group ${RESOURCE_GROUP} --query "[?ends_with(name, 'sa')].name" -o tsv)
$ GALLERY_NAME=$(az sig list --resource-group ${RESOURCE_GROUP} --query "[].name" -o tsv)
----

. Get the `arm64` {op-system} VHD URL from the cluster's boot images configuration by running the following commands:
+
[source,terminal]
----
$ VHD_URL=$(oc -n openshift-machine-config-operator get configmap/coreos-bootimages \
  -o jsonpath='{.data.stream}' | jq -r \
  '.architectures.aarch64."rhel-coreos-extensions"."azure-disk".url')
$ VHD_RELEASE=$(oc -n openshift-machine-config-operator get configmap/coreos-bootimages \
  -o jsonpath='{.data.stream}' | jq -r \
  '.architectures.aarch64."rhel-coreos-extensions"."azure-disk".release')
$ BLOB_NAME="rhcos-${VHD_RELEASE}-azure.aarch64.vhd"

> ----

Reviewer: Some steps say "Set environment variables" and some don't. Ideally, make all the env var steps use the same phrasing. As with the previous comment, we cannot have more than one command per block. Break these blocks up.

. Copy the `arm64` {op-system} VHD to the cluster's storage account by running the following commands:
+
[source,terminal]
----
$ ACCOUNT_KEY=$(az storage account keys list \
  -g ${RESOURCE_GROUP} \
  --account-name ${STORAGE_ACCOUNT_NAME} \
  --query "[0].value" -o tsv)
$ az storage blob copy start \
  --account-name ${STORAGE_ACCOUNT_NAME} \
  --account-key "${ACCOUNT_KEY}" \
  --source-uri "${VHD_URL}" \
  --destination-blob "${BLOB_NAME}" \
  --destination-container vhd

> ----

Reviewer: This looks like two steps: set an env var (`$ ACCOUNT_KEY=`) and then "Copy the `arm64` {op-system} VHD to the cluster's storage account".

I would break the first command out as another env var step and then do this as a discrete step:

```suggestion
. Copy the `arm64` {op-system} VHD to the cluster's storage account by running the following commands:
+
[source,terminal]
----
$ az storage blob copy start \
  --account-name ${STORAGE_ACCOUNT_NAME} \
  --account-key "${ACCOUNT_KEY}" \
  --source-uri "${VHD_URL}" \
  --destination-blob "${BLOB_NAME}" \
  --destination-container vhd
----
```

+
[source,terminal]
----
$ IMAGE_DEFINITION_NAME=$(az sig image-definition list \
  --resource-group ${RESOURCE_GROUP} \
  --gallery-name ${GALLERY_NAME} \
  --query "[?contains(name,'-gen2')].name" -o tsv \
  | sed 's/-gen2/-aarch64-gen2/')
$ az sig image-definition create \
  --resource-group ${RESOURCE_GROUP} \
  --gallery-name ${GALLERY_NAME} \
  --gallery-image-definition ${IMAGE_DEFINITION_NAME} \
  --publisher RedHat-gen2 \
  --offer rhcos-aarch64-gen2 \
  --sku gen2 \
  --os-type Linux \
  --architecture Arm64 \
  --hyper-v-generation V2 \
  -l ${REGION}
----

. Create a gallery image version from the copied VHD by running the following commands:
+
[source,terminal]
----
$ RHCOS_VHD_URL=$(az storage blob url \
  --account-name ${STORAGE_ACCOUNT_NAME} \
  -c vhd -n "${BLOB_NAME}" -o tsv)
$ az sig image-version create \
  --resource-group ${RESOURCE_GROUP} \
  --gallery-name ${GALLERY_NAME} \
  --gallery-image-definition ${IMAGE_DEFINITION_NAME} \
  --gallery-image-version 1.0.0 \
  --os-vhd-storage-account ${STORAGE_ACCOUNT_NAME} \
  --os-vhd-uri ${RHCOS_VHD_URL} \
  -l ${REGION}

> ----

Reviewer: Again, these are two steps - set an env var then use it. At minimum they need to be two separate blocks. You could maybe keep them as one step by adding something about the env var to the text (like so)

```suggestion
. Create an `arm64` image definition environment variable and add it to the cluster's shared image gallery by running the following commands:
+
[source,terminal]
----
$ IMAGE_DEFINITION_NAME=$(az sig image-definition list \
  --resource-group ${RESOURCE_GROUP} \
  --gallery-name ${GALLERY_NAME} \
  --query "[?contains(name,'-gen2')].name" -o tsv \
  | sed 's/-gen2/-aarch64-gen2/')
$ az sig image-definition create \
  --resource-group ${RESOURCE_GROUP} \
  --gallery-name ${GALLERY_NAME} \
  --gallery-image-definition ${IMAGE_DEFINITION_NAME} \
  --publisher RedHat-gen2 \
  --offer rhcos-aarch64-gen2 \
  --sku gen2 \
  --os-type Linux \
  --architecture Arm64 \
  --hyper-v-generation V2 \
  -l ${REGION}
----

. Create a gallery image version environment variable from the copied VHD and then create a gallery image version by running the following commands:
+
[source,terminal]
----
$ RHCOS_VHD_URL=$(az storage blob url \
  --account-name ${STORAGE_ACCOUNT_NAME} \
  -c vhd -n "${BLOB_NAME}" -o tsv)
$ az sig image-version create \
  --resource-group ${RESOURCE_GROUP} \
  --gallery-name ${GALLERY_NAME} \
  --gallery-image-definition ${IMAGE_DEFINITION_NAME} \
  --gallery-image-version 1.0.0 \
  --os-vhd-storage-account ${STORAGE_ACCOUNT_NAME} \
  --os-vhd-uri ${RHCOS_VHD_URL} \
  -l ${REGION}
----
```

But really those are both two steps and it would be better to split each up

+
[source,terminal]
----
$ RESOURCE_ID="/$(az sig image-version show \
  --resource-group ${RESOURCE_GROUP} \
  --gallery-name ${GALLERY_NAME} \
  --gallery-image-definition ${IMAGE_DEFINITION_NAME} \
  --gallery-image-version 1.0.0 \
  --query id -o tsv | cut -d'/' -f4-)"
----

. Update the control plane machine set to use the `arm64` image and an ARM-compatible VM size by running the following command:
+
[source,terminal]
----
$ oc patch controlplanemachineset.machine.openshift.io cluster \
  --type=json \
  -p "[
    {\"op\": \"replace\", \"path\": \"/spec/template/machines_v1beta1_machine_openshift_io/spec/providerSpec/value/image/resourceID\", \"value\": \"${RESOURCE_ID}\"},
    {\"op\": \"replace\", \"path\": \"/spec/template/machines_v1beta1_machine_openshift_io/spec/providerSpec/value/vmSize\", \"value\": \"<arm64_vm_size>\"}
  ]" \
  -n openshift-machine-api
----
+
Replace `<arm64_vm_size>` with an ARM-compatible VM size, such as `Standard_D8ps_v6`. The `ps` suffix in {azure-short} VM sizes denotes Ampere Arm-based processors. For information about supported instance types, see "Tested instance types for {azure-short} on 64-bit ARM infrastructures".
+
[NOTE]
====

> The control plane machine set replaces master nodes one at a time. The full rollout typically takes approximately 55 minutes for a 3-node control plane. During the rollout, `etcd` may report transient `Degraded` or `Progressing` conditions, which resolve after all masters are replaced.

Reviewer: ```suggestion
The control plane machine set replaces control plane nodes one at a time. The full rollout typically takes approximately 55 minutes for a 3-node control plane. During the rollout, `etcd` may report transient `Degraded` or `Progressing` conditions, which resolve after all control plane nodes are replaced.
```

Reviewer: Having written the majority control plane machine set content, I am not sure about this as a universal statement. My understanding is that updates to the control plane depend on the configured update strategy of the control plane machine set. I have often used this phrasing in control plane machine set content:

```
* For clusters that use the default `RollingUpdate` update strategy, the control plane machine set propagates changes to your control plane configuration automatically.

For clusters that are configured to use the `OnDelete` update strategy, you must replace your control plane machines manually.
```

====
+
[NOTE]
====
If a replacement machine fails with an "instance missing" error, delete the failed machine to allow the control plane machine set to retry the replacement.

> ====

Reviewer: We can't stack admonition blocks. I would suggest leaving this one as a note and making the previous one above it plain text instead of a note

.Verification

> * Verify that the control plane nodes are now running on the `arm64` architecture by running the following command:

Reviewer: ```suggestion
* Verify that the control plane nodes are running on the `arm64` architecture by running the following command:
```
